File: BackUpForABook/UIControls/Graph_framework/Graph/Graph/Common/DSCircle.m
//
//  DSCircle.m
//  ATreeRingMap
//
//  Created by hui wang on 11/29/15.
//  Copyright (c) 2015 AnnieLyticx. All rights reserved.
//

#import "DSCircle.h"

@implementation DSCircle
{
    @private
    
    // circel edge
    float _width;
}

-(id)init
{
    self = [super init];
    
    _size = 24;
    _show = TRUE;
    _width = 5.4;
    
    _colorLo = [UIColor colorWithRed: 0.004 green: 0.004 blue: 0.004 alpha: 0.5];
    _colorHi = [UIColor colorWithRed: 0.004 green: 0.004 blue: 0.004 alpha: 0];
    
    _outsideColor = [UIColor colorWithRed: 0.902 green: 0.902 blue: 0.898 alpha: 1];
    _insideColor = [UIColor colorWithRed: 1 green: 1 blue: 1 alpha: 1];
    
    _handle_width = 4;
    _handle_Color = [UIColor colorWithRed: 0.945 green: 0.361 blue: 0.153 alpha: 1];
    
    return self;
}

// paint the circle at given location
-(void)paint:(CGContextRef)ctx
      origin:(CGPoint)origin
{
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    
    CGFloat sVGID_Locations[] = {0, 1};
    CGGradientRef sVGID = CGGradientCreateWithColors(colorSpace, (__bridge CFArrayRef)@[(id)_colorHi.CGColor, (id)_colorLo.CGColor], sVGID_Locations);

    CGRect circleRect = CGRectMake(origin.x - _size / 2, origin.y - _size / 2, _size, _size);
    
    CGRect circleInner = CGRectMake(origin.x - (_size - _width) / 2, origin.y - (_size - _width) / 2, _size - _width, _size - _width);
    
    // Oval fill
    UIBezierPath* oval2Path = [UIBezierPath bezierPathWithOvalInRect: circleRect];
    [_insideColor setFill];
    [oval2Path fill];
    
    // Oval stroke
    UIBezierPath* oval3Path = [UIBezierPath bezierPathWithOvalInRect: circleRect];
    [_outsideColor setStroke];
    oval3Path.lineWidth = 1.16;
    [oval3Path stroke];

    UIBezierPath* oval4Path = [UIBezierPath bezierPathWithOvalInRect: circleInner];
    
    CGContextSaveGState(ctx);
    [oval4Path addClip];
    CGContextDrawLinearGradient(ctx, sVGID,
                                CGPointMake(origin.x, origin.y),
                                CGPointMake(origin.x, origin.y - (_size - _width) / 2),
                                kCGGradientDrawsBeforeStartLocation | kCGGradientDrawsAfterEndLocation);
    CGContextRestoreGState(ctx);
    
    CGGradientRelease(sVGID);
    CGColorSpaceRelease(colorSpace);
}

// paint the circle woth handle bar
-(void)paint:(CGContextRef)ctx
      origin:(CGPoint)origin
  handle_pos:(CGPoint)handle_pos
{
    // paint the handle first
    
    // outer path
    UIBezierPath* bezierOuterPath = UIBezierPath.bezierPath;
    [bezierOuterPath moveToPoint: handle_pos];
    [bezierOuterPath addLineToPoint: origin];
    
    [[UIColor whiteColor] setStroke];
    bezierOuterPath.lineWidth = _handle_width + 2.0;
    [bezierOuterPath stroke];
    
    // inner path
    UIBezierPath* bezierInnerPath = UIBezierPath.bezierPath;
    [bezierInnerPath moveToPoint: handle_pos];
    [bezierInnerPath addLineToPoint: origin];
    
    [_handle_Color setStroke];
    bezierInnerPath.lineWidth = _handle_width;
    [bezierInnerPath stroke];

    // then the circle
    [self paint:ctx origin:origin];
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/DSCircle.swift
//
//  DSCircle.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/16.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

import UIKit

class DSCircle: NSObject {
    var show: Bool?
    var size: Float?
    var colorHi: UIColor?
    var colorLo: UIColor?
    var outsideColor: UIColor?
    var insideColor: UIColor?
    var handleWidth: Float?
    var handleColor: UIColor?
    var width: Float?
    
    override init() {
        super.init()
        size = 24
        show = true
        width = 5.4
        colorLo = UIColor(red: 0.004, green: 0.004, blue: 0.004, alpha: 0.5)
        colorHi = UIColor(red: 0.004, green: 0.004, blue: 0.004, alpha: 0)
        outsideColor = UIColor(red: 0.902, green: 0.902, blue: 0.898, alpha: 1)
        insideColor = UIColor(red: 1, green: 1, blue: 1, alpha: 1)
        handleWidth = 4
        handleColor = UIColor(red: 0.945, green: 0.361, blue: 0.153, alpha: 1)
    }
    
    // paint the circle 
    func paint(_ ctx: CGContext, origin: CGPoint, handle_pos: CGPoint){
        // outer path
        let bezierOuterPath = UIBezierPath()
        bezierOuterPath.move(to: handle_pos)
        bezierOuterPath.addLine(to: origin)
        
        UIColor.white.setStroke()
        bezierOuterPath.lineWidth = CGFloat(handleWidth!) + 2.0
        bezierOuterPath.stroke()
        
        // inner path
        let bezierInnerPath = UIBezierPath()
        bezierInnerPath.move(to: handle_pos)
        bezierInnerPath.addLine(to: origin)
        
        handleColor!.setStroke()
        bezierInnerPath.lineWidth = CGFloat(handleWidth!)
        bezierInnerPath.stroke()
        
        // then the circle
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let sVGID_Locations: [CGFloat] = [0,1]
        let sVGID = CGGradient(colorsSpace: colorSpace, colors: [colorHi!.cgColor,colorLo!.cgColor] as CFArray, locations: sVGID_Locations)
        
        let circleRect = CGRect(x: origin.x - CGFloat(size! / 2.0), y: origin.y - CGFloat(size! / 2.0), width: CGFloat(size!), height: CGFloat(size!))
        let circleInner = CGRect(x: origin.x - CGFloat((size! - width!) / 2.0), y: origin.y - CGFloat((size! - width!) / 2.0), width: CGFloat(size! - width!), height: CGFloat(size! - width!))
        let oval2Path = UIBezierPath(ovalIn: circleRect)
        insideColor!.setFill()
        oval2Path.fill()
        
        let oval3Path = UIBezierPath(ovalIn: circleRect)
        outsideColor!.setStroke()
        oval3Path.lineWidth = 1.16
        oval3Path.stroke()
        
        let oval4Path = UIBezierPath(ovalIn: circleInner)
        ctx.saveGState()
        oval4Path.addClip()
        ctx.drawLinearGradient(sVGID!, start: CGPoint.init(x: origin.x, y: origin.y), end: CGPoint.init(x: origin.x, y: origin.y - CGFloat(size! - width!) / 2), options: .drawsBeforeStartLocation)
        ctx.restoreGState()
    }

}
